src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,492 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+    };
+
+    explicit fetch_worker(shard_local_fetch_context ctx)
+      : _ctx(std::move(ctx)) {}
+
+    struct worker_result {
+        std::vector<read_result> read_results;
+        // The total amount of bytes read across all results in `read_results`.
+        size_t total_size;
+    };
+
+    ss::future<worker_result> run() {
+        // Set the worker's abort source to be a child of the orchestrator's
+        // abort_source.
+        auto sub_opt = _ctx.as.subscribe([this]() noexcept {
+            _as.request_abort();
+            _completed_waiter_count.signal();
+        });
+
+        if (!sub_opt) {
+            // `_ctx.as` must've already been aborted. Query partitions once
+            // then return.
+            _as.request_abort();
+        }
+
+        auto result = co_await do_run();
+
+        // Ensure all offset waiters have returned before returning.
+        _as.request_abort();
+        co_await _waiter_gate.close();
+
+        co_return std::move(result);
+    }
+
+private:
+    shard_local_fetch_context _ctx;
+    ss::gate _waiter_gate;
+    ss::abort_source _as;
+
+    // This should roughly correspond to `_request_indexes.size()`. It's used to
+    // wait for partitions to have an offset change so that it's corresponding
+    // request in the fetch can be re-queried.
+    ss::semaphore _completed_waiter_count{0};
+    // This contains indexes into `_ctx.requests` and is populated by raft
+    // consensus waiters when an offset increases for a partition.
+    std::vector<size_t> _request_indexes;
+    // Contains the last observed value of `consensus->last_visible_index()` for
+    // every request in `_ctx.requests`. Its used to register waiters with
+    // `consensus->visible_offset_monitor()`.
+    std::vector<model::offset> _last_visible_indexes;
+
+    struct query_results {
+        std::vector<model::offset> last_visible_indexes;
+        // Indicates if any `read_result` in `results` has an error.
+        bool has_error;
+        std::vector<read_result> results;
+        size_t total_size;
+    };
+
+    ss::future<query_results>
+    query_requests(std::vector<ntp_fetch_config> requests) {
+        // The last visible indexes need to be populated before partitions
+        // are read. If they are populated afterwards then the
+        // last_visible_index could be updated after the partition is read,
+        // but before the index is queried. This leads to a race condition
+        // where we're waiting for the last_visible_index to change even
+        // though the partition has data we haven't read.
+        //
+        // Note that `last_visible_index` is used instead of the
+        // `commit_offset` as the commit offset won't change when partitions
+        // are produced to without acks=all. The `last_visible_index` more
+        // closely corresponds to the Kafka high watermark as well.
+        std::vector<model::offset> last_visible_indexes{requests.size()};
```

Comment:
Does this create a vector with `requests.size()` elements, default-initialized, or does it create a 1-element vector whose value is `requests.size()` converted to an offset?

Reply:
According to clangd it's requests.size() elements. But yeah, its confusing I agree. Will switch to ```        std::vector<model::offset> last_visible_indexes(requests.size());```

Reply:
Right, I guess because model::offset can't be initalized with a size_t but that's pretty lucky!